simulation/agent/ParenthoodAgent.java

```diff
@@ -0,0 +1,90 @@
+import com.vaticle.typedb.benchmark.common.concept.Parenthood;
+import com.vaticle.typedb.benchmark.common.concept.Person;
+import com.vaticle.typedb.benchmark.common.params.Context;
+import com.vaticle.typedb.benchmark.common.seed.RandomSource;
+import com.vaticle.typedb.benchmark.simulation.driver.Client;
+import com.vaticle.typedb.benchmark.simulation.driver.Session;
+import com.vaticle.typedb.benchmark.simulation.driver.Transaction;
+import com.vaticle.typedb.common.collection.Pair;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static com.vaticle.typedb.common.collection.Collections.list;
+import static java.util.Comparator.comparing;
+import static java.util.stream.Collectors.toCollection;
+
+public abstract class ParenthoodAgent<TX extends Transaction> extends Agent<Country, TX> {
+
+    protected ParenthoodAgent(Client<?, TX> client, Context context) {
+        super(client, context);
+    }
+
+    @Override
+    protected Class<? extends Agent> agentClass() {
+        return ParenthoodAgent.class;
+    }
+
+    @Override
+    protected List<Country> regions() {
+        return context.seedData().countries();
+    }
+
+    @Override
+    protected List<Report> run(Session<TX> session, Country country, RandomSource random) {
+        List<Report> reports = new ArrayList<>();
+        try (TX tx = session.writeTransaction()) {
+            LocalDateTime marriageDate = context.today().minusYears(context.yearsBeforeParenthood());
+            List<Marriage> marriages = matchMarriages(tx, country, marriageDate)
+                    .sorted(comparing(Marriage::licence)).collect(toCollection(ArrayList::new));
+            List<Person> newBorns = matchNewborns(tx, country, context.today())
+                    .sorted(comparing(Person::email)).collect(toCollection(ArrayList::new));
+            List<Pair<Marriage, Person>> parenthoods = random.randomAllocation(marriages, newBorns);
+
+            parenthoods.forEach(parenthood -> {
+                Optional<Parenthood> inserted = insertParenthood(tx, parenthood.first().wife().email(),
+                                                                 parenthood.first().husband().email(),
+                                                                 parenthood.second().email());
+                if (context.isReporting()) {
+                    assert inserted.isPresent();
+                    reports.add(new Report(list(parenthood.first().wife().email(), parenthood.first().husband().email(),
+                                                parenthood.second().email()),
+                                           list(inserted.get().mother().email(), inserted.get().father().email(),
+                                                inserted.get().child().email())));
+                } else assert inserted.isEmpty();
+            });
+            tx.commit();
```

Comment:
Formatting of this method can be made more readable by introducing variables in a few places.